accepted/2022/enhance-vector-codegen.md

```diff
@@ -0,0 +1,679 @@
+            VectorMask<byte> cmpCh1 = Vector.Equals(ch1, Vector.LoadUnsafe(ref searchSpace, (nuint)offset));
+            VectorMask<byte> cmpAnd = (cmpCh1 & cmpCh2);
+
+            // Early out: cmpAnd is all zeros
+            if (cmpAnd != VectorMask<byte>.Zero)
+            {
+                goto CANDIDATE_FOUND;
+            }
+
+        LOOP_FOOTER:
+            offset += Vector<byte>.Count;
+
+            if (offset == searchSpaceMinusValueTailLength)
+                return -1;
+
+            // Overlap with the current chunk for trailing elements
+            if (offset > searchSpaceMinusValueTailLengthAndVector)
+                offset = searchSpaceMinusValueTailLengthAndVector;
+
+            continue;
+
+        CANDIDATE_FOUND:
+            do
+            {
+                int bitPos = cmpAnd.FirstIndexOf(true); 
+                if (valueLength == 2 || 
+                    SequenceEqual(
+                        ref Unsafe.Add(ref searchSpace, offset + bitPos),
+                        ref value, (nuint)(uint)valueLength)) 
+                {
+                    return (int)(offset + bitPos);
+                }
+                cmpAnd = cmpAnd.SetElementCond(bitPos, false);
+            } while (mask != VectorMask<byte>.Zero);
+            goto LOOP_FOOTER;
+
+        } while (true);
+    }
+}
+
+```
+
+Note the presence of the new `VectorMask<T>` methods allows to use `Vector<T>` in place of the explicit `Vector128`, `Vector256` code path checks. When combined with `#[Vectorize]`, the JIT will probe the proper varaibles, i.e., `searchSpaceMinusValueTailLength` and compare with the threshold check `searchSpaceMinusValueTailLength - Vector<byte>.Count >= 0` at runtime to determine which `Vector<T>` length should be selected.
+## Requirements
+
+### Goals
+
+1. `Vector<T>` should allow to generate code that reaches performance within some threshold of hand-optimized intrinsics. 
+
+2. In combination with the JIT, `Vector<T>` written code should allow to adapt to the best performance of underlying platform, i.e., the JIT may generate multiple codepaths and thresholds that select at runtime which SIMD ISA to use, be it through static templated expansion or dynamic method recompilation.
+
+3. `Vector<T>` API surface should include sufficiently expressive methods to reach the first goal above. These methods allow to use `Vector<T>` generically without explicit knowledge of vector length or intrinsics.
+
+### Non-Goals
+
+
+## Stakeholders and Reviewers
+
+<!--
+We noticed that even in the cases where we have specs, we sometimes surprise key
+stakeholders because we didn't pro-actively involve them in the initial reviews
+and early design process.
+
+Please take a moment and add a bullet point list of teams and individuals you
+think should be involved in the design process and ensure they are involved
+(which might mean being tagged on GitHub issues, invited to meetings, or sent
+early drafts).
```

Comment:
Someone from Arm should look at this. Support for Arm SVE would require a Vector like interface, so it'd be good to ensure the design is suitable if SVE support gets implemented. I suspect most concerns will be around the mask implementation, and I see that Tanner has some comments below too. I'll see who we've got available to take a look....